File: go/binary_heap/heap_up_test.go
package binary_heap

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestHeapUp_RandomCases(t *testing.T) {
	for range 1_000 {
		td := newRandomTestData()

		assert.True(t, minHeapIsOK(td.heap))

		td.heap[td.randomIndex] -= td.randomValue
		actual := append([]int{}, td.heap...)
		HeapUp(actual, td.randomIndex)

		if !minHeapIsOK(actual) {
			assert.FailNowf(t, "",
				"HeapUp(%v, %v): not a valid heap after the call:\n%v",
				td.heap, td.randomIndex, actual,
			)
		}

		if !haveSameElements(actual, td.heap) {
			assert.FailNowf(t, "",
				"HeapUp(%v, %v): not same elements after the call:\n%v",
				td.heap, td.randomIndex, actual,
			)
		}
	}
}

File: python/binary_heap/heap_up_test.py
from heap_up import heap_up
from bh_random_test_data import RandomTestData
from bh_random_test_data import min_heap_is_ok


def test_heap_up_random_cases():
	for _ in range(1_000):
		td = RandomTestData()

		assert min_heap_is_ok(td.heap)

		td.heap[td.random_index] -= td.randomValue

		got = td.heap.copy()
		heap_up(got, td.random_index)

		assert min_heap_is_ok(got), \
			f'heap_up({td.heap}, {td.random_index}): not a valid heap after the call:\n{got}'

		assert sorted(got) == sorted(td.heap), \
			f'HeapUp({td.heap}, {td.random_index}): not same elements after the call:\n{got}'
